debug shape guards

avikchaudhuri opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #95848





Adds logging when shape guards are added and when symbols are specialized to constants.

Differential Revision: [D43719743](https://our.internmc.facebook.com/intern/diff/D43719743/)

Differential Revision: [D43719743](https://our.internmc.facebook.com/intern/diff/D43719743)

cc @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

avikchaudhuri committed:
debug shape guards



Differential Revision: [D43719743](https://our.internmc.facebook.com/intern/diff/D43719743/)

[ghstack-poisoned]

avikchaudhuri commented:
Example: 
```
torch._dynamo.config.log_level = logging.WARNING
torch._dynamo.config.print_shape_guards = True
torch._dynamo.config.print_specializations = True

def foo(x, y, z):
    if (
        2 * x.shape[0] == y.shape[0] + z.shape[0]
        and x.shape[0] == z.shape[0] + 1
        and 2 * z.shape[0] == y.shape[0] + 4
    ):
        return 2 * y
    else:
        return 2 + y


m, gs = dynamo.export(
    foo,
    torch.rand(7),
    torch.rand(8),
    torch.rand(6),
    aten_graph=True,
    tracing_mode="symbolic",
)

```

This code adds 3 guards, which are also sufficient to specialize the shapes of all 3 variables. Log:

```
[2023-03-01 17:02:41,988] torch._dynamo.guards: [WARNING] Adding shape guard Eq(2*s0, s1 + s2):
  File "<string>", line 38, in <module>
  File "<string>", line 36, in __run
  File "/usr/local/fbcode/platform010/lib/python3.8/runpy.py", line 194, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/local/fbcode/platform010/lib/python3.8/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "<snip>/scripts/avik/pt2/example.py", line 371, in <module>
    m, gs = dynamo.export(
  File "<snip>/torch/_dynamo/eval_frame.py", line 607, in export
    result_traced = opt_f(*args, **kwargs)
  File "<snip>/torch/_dynamo/eval_frame.py", line 215, in _fn
    return fn(*args, **kwargs)
  File "<snip>/torch/_dynamo/eval_frame.py", line 343, in catch_errors
    return callback(frame, cache_size, hooks)
  File "<snip>/torch/_dynamo/convert_frame.py", line 104, in _fn
    return fn(*args, **kwargs)
  File "<snip>/torch/_dynamo/convert_frame.py", line 262, in _convert_frame_assert
    return _compile(
  File "<snip>/torch/_dynamo/utils.py", line 164, in time_wrapper
    r = func(*args, **kwargs)
  File "<snip>/torch/_dynamo/convert_frame.py", line 324, in _compile
    out_code = transform_code_object(code, transform)
  File "<snip>/torch/_dynamo/bytecode_transformation.py", line 530, in transform_code_object
    transformations(instructions, code_options)
  File "<snip>/torch/_dynamo/convert_frame.py", line 311, in transform
    tracer.run()
  File "<snip>/torch/_dynamo/symbolic_convert.py", line 1862, in run
    super().run()
  File "<snip>/torch/_dynamo/symbolic_convert.py", line 619, in run
    and self.step()
  File "<snip>/torch/_dynamo/symbolic_convert.py", line 583, in step
    getattr(self, inst.opname)(inst)
  File "<snip>/torch/_dynamo/symbolic_convert.py", line 329, in inner
    eval_result = value.evaluate_expr(self.output)
  File "<snip>/torch/_dynamo/variables/tensor.py", line 464, in evaluate_expr
    return guard_scalar(self.sym_num)
  File "<snip>/torch/fx/experimental/symbolic_shapes.py", line 188, in guard_scalar
    return guard_bool(a)
  File "<snip>/torch/fx/experimental/symbolic_shapes.py", line 246, in guard_bool
    return a.node.guard_bool("", 0)  # NB: uses Python backtrace
  File "<snip>/torch/fx/experimental/symbolic_shapes.py", line 541, in guard_bool
    r = self.shape_env.evaluate_expr(self.expr, self.hint)
  File "<snip>/torch/fx/experimental/symbolic_shapes.py", line 1911, in evaluate_expr
    self.add_guard(expr)

[2023-03-01 17:02:42,010] torch._dynamo.guards: [WARNING] Adding shape guard Eq(s0, s2 + 1):
  File "<string>", line 38, in <module>
  File "<string>", line 36, in __run
  File "/usr/local/fbcode/platform010/lib/python3.8/runpy.py", line 194, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/local/fbcode/platform010/lib/python3.8/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "<snip>/scripts/avik/pt2/example.py", line 371, in <module>
    m, gs = dynamo.export(
  File "<snip>/torch/_dynamo/eval_frame.py", line 607, in export
    result_traced = opt_f(*args, **kwargs)
  File "<snip>/torch/_dynamo/eval_frame.py", line 215, in _fn
    return fn(*args, **kwargs)
  File "<snip>/torch/_dynamo/eval_frame.py", line 343, in catch_errors
    return callback(frame, cache_size, hooks)
  File "<snip>/torch/_dynamo/convert_frame.py", line 104, in _fn
    return fn(*args, **kwargs)
  File "<snip>/torch/_dynamo/convert_frame.py", line 262, in _convert_frame_assert
    return _compile(
  File "<snip>/torch/_dynamo/utils.py", line 164, in time_wrapper
    r = func(*args, **kwargs)
  File "<snip>/torch/_dynamo/convert_frame.py", line 324, in _compile
    out_code = transform_code_object(code, transform)
  File "<snip>/torch/_dynamo/bytecode_transformation.py", line 530, in transform_code_object
    transformations(instructions, code_options)
  File "<snip>/torch/_dynamo/convert_frame.py", line 311, in transform
    tracer.run()
  File "<snip>/torch/_dynamo/symbolic_convert.py", line 1862, in run
    super().run()
  File "<snip>/torch/_dynamo/symbolic_convert.py", line 619, in run
    and self.step()
  File "<snip>/torch/_dynamo/symbolic_convert.py", line 583, in step
    getattr(self, inst.opname)(inst)
  File "<snip>/torch/_dynamo/symbolic_convert.py", line 329, in inner
    eval_result = value.evaluate_expr(self.output)
  File "<snip>/torch/_dynamo/variables/tensor.py", line 464, in evaluate_expr
    return guard_scalar(self.sym_num)
  File "<snip>/torch/fx/experimental/symbolic_shapes.py", line 188, in guard_scalar
    return guard_bool(a)
  File "<snip>/torch/fx/experimental/symbolic_shapes.py", line 246, in guard_bool
    return a.node.guard_bool("", 0)  # NB: uses Python backtrace
  File "<snip>/torch/fx/experimental/symbolic_shapes.py", line 541, in guard_bool
    r = self.shape_env.evaluate_expr(self.expr, self.hint)
  File "<snip>/torch/fx/experimental/symbolic_shapes.py", line 1911, in evaluate_expr
    self.add_guard(expr)

[2023-03-01 17:02:42,026] torch._dynamo.guards: [WARNING] Specializing s2 to 6:
  File "<string>", line 38, in <module>
  File "<string>", line 36, in __run
  File "/usr/local/fbcode/platform010/lib/python3.8/runpy.py", line 194, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/local/fbcode/platform010/lib/python3.8/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "<snip>/scripts/avik/pt2/example.py", line 371, in <module>
    m, gs = dynamo.export(
  File "<snip>/torch/_dynamo/eval_frame.py", line 607, in export
    result_traced = opt_f(*args, **kwargs)
  File "<snip>/torch/_dynamo/eval_frame.py", line 215, in _fn
    return fn(*args, **kwargs)
  File "<snip>/torch/_dynamo/eval_frame.py", line 343, in catch_errors
    return callback(frame, cache_size, hooks)
  File "<snip>/torch/_dynamo/convert_frame.py", line 104, in _fn
    return fn(*args, **kwargs)
  File "<snip>/torch/_dynamo/convert_frame.py", line 262, in _convert_frame_assert
    return _compile(
  File "<snip>/torch/_dynamo/utils.py", line 164, in time_wrapper
    r = func(*args, **kwargs)
  File "<snip>/torch/_dynamo/convert_frame.py", line 324, in _compile
    out_code = transform_code_object(code, transform)
  File "<snip>/torch/_dynamo/bytecode_transformation.py", line 530, in transform_code_object
    transformations(instructions, code_options)
  File "<snip>/torch/_dynamo/convert_frame.py", line 311, in transform
    tracer.run()
  File "<snip>/torch/_dynamo/symbolic_convert.py", line 1862, in run
    super().run()
  File "<snip>/torch/_dynamo/symbolic_convert.py", line 619, in run
    and self.step()
  File "<snip>/torch/_dynamo/symbolic_convert.py", line 583, in step
    getattr(self, inst.opname)(inst)
  File "<snip>/torch/_dynamo/symbolic_convert.py", line 329, in inner
    eval_result = value.evaluate_expr(self.output)
  File "<snip>/torch/_dynamo/variables/tensor.py", line 464, in evaluate_expr
    return guard_scalar(self.sym_num)
  File "<snip>/torch/fx/experimental/symbolic_shapes.py", line 188, in guard_scalar
    return guard_bool(a)
  File "<snip>/torch/fx/experimental/symbolic_shapes.py", line 246, in guard_bool
    return a.node.guard_bool("", 0)  # NB: uses Python backtrace
  File "<snip>/torch/fx/experimental/symbolic_shapes.py", line 541, in guard_bool
    r = self.shape_env.evaluate_expr(self.expr, self.hint)
  File "<snip>/torch/fx/experimental/symbolic_shapes.py", line 1903, in evaluate_expr
    self._maybe_guard_eq(expr, bool(concrete_val))
  File "<snip>/torch/fx/experimental/symbolic_shapes.py", line 1849, in _maybe_guard_eq
    self._set_replacement(cast(sympy.Symbol, free[0]), new_var)

[2023-03-01 17:02:42,027] torch._dynamo.guards: [WARNING] Adding shape guard Eq(2*s2, s2 + 6):
  File "<string>", line 38, in <module>
  File "<string>", line 36, in __run
  File "/usr/local/fbcode/platform010/lib/python3.8/runpy.py", line 194, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/local/fbcode/platform010/lib/python3.8/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "<snip>/scripts/avik/pt2/example.py", line 371, in <module>
    m, gs = dynamo.export(
  File "<snip>/torch/_dynamo/eval_frame.py", line 607, in export
    result_traced = opt_f(*args, **kwargs)
  File "<snip>/torch/_dynamo/eval_frame.py", line 215, in _fn
    return fn(*args, **kwargs)
  File "<snip>/torch/_dynamo/eval_frame.py", line 343, in catch_errors
    return callback(frame, cache_size, hooks)
  File "<snip>/torch/_dynamo/convert_frame.py", line 104, in _fn
    return fn(*args, **kwargs)
  File "<snip>/torch/_dynamo/convert_frame.py", line 262, in _convert_frame_assert
    return _compile(
  File "<snip>/torch/_dynamo/utils.py", line 164, in time_wrapper
    r = func(*args, **kwargs)
  File "<snip>/torch/_dynamo/convert_frame.py", line 324, in _compile
    out_code = transform_code_object(code, transform)
  File "<snip>/torch/_dynamo/bytecode_transformation.py", line 530, in transform_code_object
    transformations(instructions, code_options)
  File "<snip>/torch/_dynamo/convert_frame.py", line 311, in transform
    tracer.run()
  File "<snip>/torch/_dynamo/symbolic_convert.py", line 1862, in run
    super().run()
  File "<snip>/torch/_dynamo/symbolic_convert.py", line 619, in run
    and self.step()
  File "<snip>/torch/_dynamo/symbolic_convert.py", line 583, in step
    getattr(self, inst.opname)(inst)
  File "<snip>/torch/_dynamo/symbolic_convert.py", line 329, in inner
    eval_result = value.evaluate_expr(self.output)
  File "<snip>/torch/_dynamo/variables/tensor.py", line 464, in evaluate_expr
    return guard_scalar(self.sym_num)
  File "<snip>/torch/fx/experimental/symbolic_shapes.py", line 188, in guard_scalar
    return guard_bool(a)
  File "<snip>/torch/fx/experimental/symbolic_shapes.py", line 246, in guard_bool
    return a.node.guard_bool("", 0)  # NB: uses Python backtrace
  File "<snip>/torch/fx/experimental/symbolic_shapes.py", line 541, in guard_bool
    r = self.shape_env.evaluate_expr(self.expr, self.hint)
  File "<snip>/torch/fx/experimental/symbolic_shapes.py", line 1911, in evaluate_expr
    self.add_guard(expr)

[2023-03-01 17:02:42,325] torch._dynamo.guards: [WARNING] Specializing s0 to 7:
  File "<string>", line 38, in <module>
  File "<string>", line 36, in __run
  File "/usr/local/fbcode/platform010/lib/python3.8/runpy.py", line 194, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/local/fbcode/platform010/lib/python3.8/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "<snip>/scripts/avik/pt2/example.py", line 371, in <module>
    m, gs = dynamo.export(
  File "<snip>/torch/_dynamo/eval_frame.py", line 607, in export
    result_traced = opt_f(*args, **kwargs)
  File "<snip>/torch/_dynamo/eval_frame.py", line 215, in _fn
    return fn(*args, **kwargs)
  File "<snip>/torch/_dynamo/eval_frame.py", line 343, in catch_errors
    return callback(frame, cache_size, hooks)
  File "<snip>/torch/_dynamo/convert_frame.py", line 104, in _fn
    return fn(*args, **kwargs)
  File "<snip>/torch/_dynamo/convert_frame.py", line 262, in _convert_frame_assert
    return _compile(
  File "<snip>/torch/_dynamo/utils.py", line 164, in time_wrapper
    r = func(*args, **kwargs)
  File "<snip>/torch/_dynamo/convert_frame.py", line 364, in _compile
    check_fn = CheckFunctionManager(
  File "<snip>/torch/_dynamo/guards.py", line 590, in __init__
    guard.create(local_builder, global_builder)
  File "<snip>/torch/_guards.py", line 158, in create
    return self.create_fn(self.source.select(local_builder, global_builder), self)
  File "<snip>/torch/_dynamo/guards.py", line 417, in SHAPE_ENV
    guards = output_graph.shape_env.produce_guards(
  File "<snip>/torch/fx/experimental/symbolic_shapes.py", line 1523, in produce_guards
    track_symint(property_source, ss)
  File "<snip>/torch/fx/experimental/symbolic_shapes.py", line 1486, in track_symint
    s = val.node.expr
  File "<snip>/torch/fx/experimental/symbolic_shapes.py", line 341, in expr
    self._update_expr()
  File "<snip>/torch/fx/experimental/symbolic_shapes.py", line 377, in _update_expr
    self._expr = self.shape_env.replace(self._expr)
  File "<snip>/torch/fx/experimental/symbolic_shapes.py", line 1150, in wrapper
    return fn_cache(self, *args, **kwargs)
  File "<snip>/torch/fx/experimental/symbolic_shapes.py", line 1705, in replace
    replacements = {s: self._find(cast(sympy.Symbol, s)) for s in expr.free_symbols}
  File "<snip>/torch/fx/experimental/symbolic_shapes.py", line 1705, in <dictcomp>
    replacements = {s: self._find(cast(sympy.Symbol, s)) for s in expr.free_symbols}
  File "<snip>/torch/fx/experimental/symbolic_shapes.py", line 1150, in wrapper
    return fn_cache(self, *args, **kwargs)
  File "<snip>/torch/fx/experimental/symbolic_shapes.py", line 1809, in _find
    self._set_replacement(a, self.replacements[a].xreplace(cur_replace))

[2023-03-01 17:02:42,326] torch._dynamo.guards: [WARNING] Specializing s1 to 8:
  File "<string>", line 38, in <module>
  File "<string>", line 36, in __run
  File "/usr/local/fbcode/platform010/lib/python3.8/runpy.py", line 194, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/local/fbcode/platform010/lib/python3.8/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "<snip>/scripts/avik/pt2/example.py", line 371, in <module>
    m, gs = dynamo.export(
  File "<snip>/torch/_dynamo/eval_frame.py", line 607, in export
    result_traced = opt_f(*args, **kwargs)
  File "<snip>/torch/_dynamo/eval_frame.py", line 215, in _fn
    return fn(*args, **kwargs)
  File "<snip>/torch/_dynamo/eval_frame.py", line 343, in catch_errors
    return callback(frame, cache_size, hooks)
  File "<snip>/torch/_dynamo/convert_frame.py", line 104, in _fn
    return fn(*args, **kwargs)
  File "<snip>/torch/_dynamo/convert_frame.py", line 262, in _convert_frame_assert
    return _compile(
  File "<snip>/torch/_dynamo/utils.py", line 164, in time_wrapper
    r = func(*args, **kwargs)
  File "<snip>/torch/_dynamo/convert_frame.py", line 364, in _compile
    check_fn = CheckFunctionManager(
  File "<snip>/torch/_dynamo/guards.py", line 590, in __init__
    guard.create(local_builder, global_builder)
  File "<snip>/torch/_guards.py", line 158, in create
    return self.create_fn(self.source.select(local_builder, global_builder), self)
  File "<snip>/torch/_dynamo/guards.py", line 417, in SHAPE_ENV
    guards = output_graph.shape_env.produce_guards(
  File "<snip>/torch/fx/experimental/symbolic_shapes.py", line 1554, in produce_guards
    if self._maybe_evaluate_static(g) is not None:
  File "<snip>/torch/fx/experimental/symbolic_shapes.py", line 1150, in wrapper
    return fn_cache(self, *args, **kwargs)
  File "<snip>/torch/fx/experimental/symbolic_shapes.py", line 1664, in _maybe_evaluate_static
    expr = self.simplify(expr)
  File "<snip>/torch/fx/experimental/symbolic_shapes.py", line 1150, in wrapper
    return fn_cache(self, *args, **kwargs)
  File "<snip>/torch/fx/experimental/symbolic_shapes.py", line 1720, in simplify
    expr = self.replace(expr)
  File "<snip>/torch/fx/experimental/symbolic_shapes.py", line 1150, in wrapper
    return fn_cache(self, *args, **kwargs)
  File "<snip>/torch/fx/experimental/symbolic_shapes.py", line 1705, in replace
    replacements = {s: self._find(cast(sympy.Symbol, s)) for s in expr.free_symbols}
  File "<snip>/torch/fx/experimental/symbolic_shapes.py", line 1705, in <dictcomp>
    replacements = {s: self._find(cast(sympy.Symbol, s)) for s in expr.free_symbols}
  File "<snip>/torch/fx/experimental/symbolic_shapes.py", line 1150, in wrapper
    return fn_cache(self, *args, **kwargs)
  File "<snip>/torch/fx/experimental/symbolic_shapes.py", line 1809, in _find
    self._set_replacement(a, self.replacements[a].xreplace(cur_replace))


```

voznesenskym commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1868,6 +1875,13 @@ def _simplify_floor_div(self, expr):
+        stack = ''.join(traceback.format_list(traceback.extract_stack()[:-1]))
+        guard = ShapeGuard(expr, stack)
+        if torch._dynamo.config.print_shape_guards:
+            torch._dynamo.guards.log.warning(f"Adding shape guard {expr}:\n{stack}")
```
We already have guard printing right? Why do we want more?

avikchaudhuri replied:
You mean at the end, right? @ezyang mentioned it would have been useful for him to see the stack trace as they are added when debugging (e.g.) https://github.com/pytorch/pytorch/issues/93468#issuecomment-1412520722

Yes, I realize it is also added to the shape guard. So we could also print guards at the end (by default we omit the stacks, I suppose that can be changed under a flag...which one?).

voznesenskym replied:
That makes sense, I would reuse the flag for dynamo guard printing then, and add a comment to the redundancy.

voznesenskym commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1786,6 +1786,13 @@ def _make_data_dependent_error(self, expr):
             # problem
         )
 
+    def _set_replacement(self, a: "sympy.Symbol", expr: "sympy.Expr"):
```
I do not think setting replacements is the same as specialization?

avikchaudhuri replied:
Right, this is just collecting all the cases that cause a replacement to be added / updated. Specialization is when they actually simplify to ints.

voznesenskym replied:
But the flag name is `print_specializations` and the log comment is too :P

avikchaudhuri replied:
To be clear, I'm only logging when `expr` is a int / float, see check right below.

avikchaudhuri commented:
As follow up, I'll also look into the ability to "monitor" a particular shape variable. Also TODO: showing actual user-code locations that trigger guards / specializations (in addition to the internal call stack within the compiler).

voznesenskym commented on torch/_dynamo/config.py:
```diff
@@ -103,6 +103,12 @@
 print_graph_breaks = False
 
+# Show a warning for every shape guard
+print_shape_guards = False
```
IMO kill this and gate this code under TORCHDYNAMO_PRINT_GUARDS?

avikchaudhuri replied:
I did as @voznesenskym asked for now.

voznesenskym commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1868,6 +1875,13 @@ def _simplify_floor_div(self, expr):
         return self.simplify(expr)
 
+    def _add_guard(self, expr: "sympy.Expr") -> None:
+        stack = ''.join(traceback.format_list(traceback.extract_stack()[:-1]))
```
this changed from `:-2` to `:-1`

ezyang replied:
It is right to renumber but you want to go to -3 not -1

avikchaudhuri replied:
I don't understand. -1 cuts only the last frame, which is the one calculating the stack; on the other hand it gives the stack up to the caller of `_add_guard` . This should be useful. It certainly is for the other place, because there are multiple different callers of `_set_replacement`. 

If anything I think we should snip from the front, because the first couple of entries are of the form `File "<string>" ...` and the next couple are from `runpy.py`.

voznesenskym commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1786,6 +1786,13 @@ def _make_data_dependent_error(self, expr):
+    def _set_replacement(self, a: "sympy.Symbol", expr: "sympy.Expr"):
+        if a not in self.replacements or expr != self.replacements[a]:
+            if torch._dynamo.config.print_specializations and isinstance(expr, (sympy.Integer, sympy.Float)):
+                stack = ''.join(traceback.format_list(traceback.extract_stack()[:-1]))
```
If you are already helper methoding everything here, pull this into one too

voznesenskym commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1786,6 +1786,13 @@ def _make_data_dependent_error(self, expr):
             # problem
         )
 
+    def _set_replacement(self, a: "sympy.Symbol", expr: "sympy.Expr"):
```
nit, your other method has a mypy annotation for returning None, this one does not

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1799,7 +1806,7 @@ def _find(self, a: "sympy.Symbol") -> "sympy.Expr":
         res = self.replacements[a]
         cur_replace = {s: self._find(s) for s in res.free_symbols}
-        self.replacements[a] = self.replacements[a].xreplace(cur_replace)
+        self._set_replacement(a, self.replacements[a].xreplace(cur_replace))
```
It's actually not obvious to me that you want to report union find path replacement the same way as new replacements discovered by unification. A union find replacement is never from the user, so I think I'm less likely to care about them.

avikchaudhuri replied:
Maybe, but IMO it is good to have both. From what I can tell, just before adding a guard we might turn it into a replacement, which is like a equation; at that point we may or may not see a direct constant replacement. These get simplified at various points (e.g., even accessing the property `.expr` of a sym node might trigger simplification), and such steps might finally cause some constant replacement to be discovered.

If we believe that the user can simplify equations in their head to know when something has turned into a constant, then we only need the user-facing info. But it's not easy to see that always.

One way to think about this is that the "Adding shape guard..." logs are high level (and sometimes coincide with immediate specialization) and the "Specializing..." fills in everything else as fallback.

ezyang commented:
The default logging seems too wordy. I would only expect to get backtraces at the very highest debug level, at a lower debug level I'd just like to see what kinds of updates we do.

Have you tried dogfooding the logging here to diagnose https://github.com/pytorch/pytorch/issues/93468 ? I would highly recommend it.

avikchaudhuri commented:
@ezyang I used this on https://github.com/pytorch/pytorch/issues/93468#issuecomment-1412520722

Output shows the specialization (it's directly from added guard):

```
[2023-03-02 09:57:04,270] torch._dynamo.guards: [WARNING] Specializing s1 to 10:
  File "<string>", line 38, in <module>
  File "<string>", line 36, in __run
  File "/usr/local/fbcode/platform010/lib/python3.8/runpy.py", line 194, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/local/fbcode/platform010/lib/python3.8/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "<snip>/scripts/avik/pt2/example.py", line 361, in <module>
    model(a)
  File "<snip>/torch/_dynamo/eval_frame.py", line 82, in __call__
    return self.dynamo_ctx(self._orig_mod.__call__)(*args, **kwargs)
  File "<snip>/torch/_dynamo/eval_frame.py", line 215, in _fn
    return fn(*args, **kwargs)
  File "<snip>/torch/nn/modules/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "<snip>/scripts/avik/pt2/example.py", line 350, in forward
    b = sequence_mask(a)
  File "<snip>/scripts/avik/pt2/example.py", line 336, in sequence_mask
    torch.arange(0, max_len, device=lengths.device)
  File "<snip>/torch/_dynamo/eval_frame.py", line 343, in catch_errors
    return callback(frame, cache_size, hooks)
  File "<snip>/torch/_dynamo/convert_frame.py", line 404, in _convert_frame
    result = inner_convert(frame, cache_size, hooks)
  File "<snip>/torch/_dynamo/convert_frame.py", line 104, in _fn
    return fn(*args, **kwargs)
  File "<snip>/torch/_dynamo/convert_frame.py", line 262, in _convert_frame_assert
    return _compile(
  File "<snip>/torch/_dynamo/utils.py", line 164, in time_wrapper
    r = func(*args, **kwargs)
  File "<snip>/torch/_dynamo/convert_frame.py", line 324, in _compile
    out_code = transform_code_object(code, transform)
  File "<snip>/torch/_dynamo/bytecode_transformation.py", line 530, in transform_code_object
    transformations(instructions, code_options)
  File "<snip>/torch/_dynamo/convert_frame.py", line 311, in transform
    tracer.run()
  File "<snip>/torch/_dynamo/symbolic_convert.py", line 1862, in run
    super().run()
  File "<snip>/torch/_dynamo/symbolic_convert.py", line 619, in run
    and self.step()
  File "<snip>/torch/_dynamo/symbolic_convert.py", line 583, in step
    getattr(self, inst.opname)(inst)
  File "<snip>/torch/_dynamo/symbolic_convert.py", line 349, in wrapper
    return inner_fn(self, inst)
  File "<snip>/torch/_dynamo/symbolic_convert.py", line 1014, in CALL_FUNCTION
    self.call_function(fn, args, {})
  File "<snip>/torch/_dynamo/symbolic_convert.py", line 517, in call_function
    self.push(fn.call_function(self, args, kwargs))
  File "<snip>/torch/_dynamo/variables/misc.py", line 759, in call_function
    return self.obj.call_method(tx, self.name, args, kwargs).add_options(self)
  File "<snip>/torch/_dynamo/variables/tensor.py", line 424, in call_method
    return wrap_fx_proxy(
  File "<snip>/torch/_dynamo/variables/builder.py", line 871, in wrap_fx_proxy
    return wrap_fx_proxy_cls(
  File "<snip>/torch/_dynamo/variables/builder.py", line 906, in wrap_fx_proxy_cls
    example_value = get_fake_value(proxy.node, tx)
  File "<snip>/torch/_dynamo/utils.py", line 1152, in get_fake_value
    return wrap_fake_exception(
  File "<snip>/torch/_dynamo/utils.py", line 808, in wrap_fake_exception
    return fn()
  File "<snip>/torch/_dynamo/utils.py", line 1153, in <lambda>
    lambda: run_node(tx.output, node, args, kwargs, nnmodule)
  File "<snip>/torch/_dynamo/utils.py", line 1196, in run_node
    return getattr(args[0], node.target)(*args[1:], **kwargs)
  File "<snip>/torch/utils/_stats.py", line 20, in wrapper
    return fn(*args, **kwargs)
  File "<snip>/torch/_subclasses/fake_tensor.py", line 1060, in __torch_dispatch__
    return self.dispatch(func, types, args, kwargs)
  File "<snip>/torch/_subclasses/fake_tensor.py", line 1213, in dispatch
    return decomposition_table[func](*args, **kwargs)
  File "<snip>/torch/_prims_common/wrappers.py", line 220, in _fn
    result = fn(*args, **kwargs)
  File "<snip>/torch/_prims_common/wrappers.py", line 130, in _fn
    result = fn(**bound.arguments)
  File "<snip>/torch/_refs/__init__.py", line 926, in _ref
    a, b = _maybe_broadcast(a, b)
  File "<snip>/torch/_refs/__init__.py", line 348, in _maybe_broadcast
    common_shape = _broadcast_shapes(
  File "<snip>/torch/_refs/__init__.py", line 336, in _broadcast_shapes
    if common_shape[idx] != shape[idx]:
  File "<snip>/torch/__init__.py", line 350, in __bool__
    return self.node.bool_()
  File "<snip>/torch/fx/experimental/symbolic_shapes.py", line 549, in bool_
    return self.guard_bool("", 0)
  File "<snip>/torch/fx/experimental/symbolic_shapes.py", line 541, in guard_bool
    r = self.shape_env.evaluate_expr(self.expr, self.hint)
  File "<snip>/torch/fx/experimental/symbolic_shapes.py", line 1903, in evaluate_expr
    self._maybe_guard_eq(expr, bool(concrete_val))
  File "<snip>/torch/fx/experimental/symbolic_shapes.py", line 1849, in _maybe_guard_eq
    self._set_replacement(cast(sympy.Symbol, free[0]), new_var)

[2023-03-02 09:57:04,272] torch._dynamo.guards: [WARNING] Adding shape guard Eq(s1, 10):
  File "<string>", line 38, in <module>
  File "<string>", line 36, in __run
  File "/usr/local/fbcode/platform010/lib/python3.8/runpy.py", line 194, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/local/fbcode/platform010/lib/python3.8/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "<snip>/scripts/avik/pt2/example.py", line 361, in <module>
    model(a)
  File "<snip>/torch/_dynamo/eval_frame.py", line 82, in __call__
    return self.dynamo_ctx(self._orig_mod.__call__)(*args, **kwargs)
  File "<snip>/torch/_dynamo/eval_frame.py", line 215, in _fn
    return fn(*args, **kwargs)
  File "<snip>/torch/nn/modules/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "<snip>/scripts/avik/pt2/example.py", line 350, in forward
    b = sequence_mask(a)
  File "<snip>/scripts/avik/pt2/example.py", line 336, in sequence_mask
    torch.arange(0, max_len, device=lengths.device)
  File "<snip>/torch/_dynamo/eval_frame.py", line 343, in catch_errors
    return callback(frame, cache_size, hooks)
  File "<snip>/torch/_dynamo/convert_frame.py", line 404, in _convert_frame
    result = inner_convert(frame, cache_size, hooks)
  File "<snip>/torch/_dynamo/convert_frame.py", line 104, in _fn
    return fn(*args, **kwargs)
  File "<snip>/torch/_dynamo/convert_frame.py", line 262, in _convert_frame_assert
    return _compile(
  File "<snip>/torch/_dynamo/utils.py", line 164, in time_wrapper
    r = func(*args, **kwargs)
  File "<snip>/torch/_dynamo/convert_frame.py", line 324, in _compile
    out_code = transform_code_object(code, transform)
  File "<snip>/torch/_dynamo/bytecode_transformation.py", line 530, in transform_code_object
    transformations(instructions, code_options)
  File "<snip>/torch/_dynamo/convert_frame.py", line 311, in transform
    tracer.run()
  File "<snip>/torch/_dynamo/symbolic_convert.py", line 1862, in run
    super().run()
  File "<snip>/torch/_dynamo/symbolic_convert.py", line 619, in run
    and self.step()
  File "<snip>/torch/_dynamo/symbolic_convert.py", line 583, in step
    getattr(self, inst.opname)(inst)
  File "<snip>/torch/_dynamo/symbolic_convert.py", line 349, in wrapper
    return inner_fn(self, inst)
  File "<snip>/torch/_dynamo/symbolic_convert.py", line 1014, in CALL_FUNCTION
    self.call_function(fn, args, {})
  File "<snip>/torch/_dynamo/symbolic_convert.py", line 517, in call_function
    self.push(fn.call_function(self, args, kwargs))
  File "<snip>/torch/_dynamo/variables/misc.py", line 759, in call_function
    return self.obj.call_method(tx, self.name, args, kwargs).add_options(self)
  File "<snip>/torch/_dynamo/variables/tensor.py", line 424, in call_method
    return wrap_fx_proxy(
  File "<snip>/torch/_dynamo/variables/builder.py", line 871, in wrap_fx_proxy
    return wrap_fx_proxy_cls(
  File "<snip>/torch/_dynamo/variables/builder.py", line 906, in wrap_fx_proxy_cls
    example_value = get_fake_value(proxy.node, tx)
  File "<snip>/torch/_dynamo/utils.py", line 1152, in get_fake_value
    return wrap_fake_exception(
  File "<snip>/torch/_dynamo/utils.py", line 808, in wrap_fake_exception
    return fn()
  File "<snip>/torch/_dynamo/utils.py", line 1153, in <lambda>
    lambda: run_node(tx.output, node, args, kwargs, nnmodule)
  File "<snip>/torch/_dynamo/utils.py", line 1196, in run_node
    return getattr(args[0], node.target)(*args[1:], **kwargs)
  File "<snip>/torch/utils/_stats.py", line 20, in wrapper
    return fn(*args, **kwargs)
  File "<snip>/torch/_subclasses/fake_tensor.py", line 1060, in __torch_dispatch__
    return self.dispatch(func, types, args, kwargs)
  File "<snip>/torch/_subclasses/fake_tensor.py", line 1213, in dispatch
    return decomposition_table[func](*args, **kwargs)
  File "<snip>/torch/_prims_common/wrappers.py", line 220, in _fn
    result = fn(*args, **kwargs)
  File "<snip>/torch/_prims_common/wrappers.py", line 130, in _fn
    result = fn(**bound.arguments)
  File "<snip>/torch/_refs/__init__.py", line 926, in _ref
    a, b = _maybe_broadcast(a, b)
  File "<snip>/torch/_refs/__init__.py", line 348, in _maybe_broadcast
    common_shape = _broadcast_shapes(
  File "<snip>/torch/_refs/__init__.py", line 336, in _broadcast_shapes
    if common_shape[idx] != shape[idx]:
  File "<snip>/torch/__init__.py", line 350, in __bool__
    return self.node.bool_()
  File "<snip>/torch/fx/experimental/symbolic_shapes.py", line 549, in bool_
    return self.guard_bool("", 0)
  File "<snip>/torch/fx/experimental/symbolic_shapes.py", line 541, in guard_bool
    r = self.shape_env.evaluate_expr(self.expr, self.hint)
  File "<snip>/torch/fx/experimental/symbolic_shapes.py", line 1913, in evaluate_expr
    self._add_guard(sympy.Not(expr))
```

avikchaudhuri commented:
To deal with verbosity, maybe I should move the backtrace printing to DEBUG and keep just the one liner in WARNING?

avikchaudhuri committed:
Update on "debug shape guards"


Differential Revision: [D43719743](https://our.internmc.facebook.com/intern/diff/D43719743/)

cc soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

avikchaudhuri commented:
Now under WARNING I only log this for the 3 guard / 3 specialization example above:
```
[2023-03-02 11:12:45,742] torch._dynamo.guards: [WARNING] Adding shape guard Eq(2*s0, s1 + s2)
[2023-03-02 11:12:45,765] torch._dynamo.guards: [WARNING] Adding shape guard Eq(s0, s2 + 1)
[2023-03-02 11:12:45,780] torch._dynamo.guards: [WARNING] Specializing s2 to 6
[2023-03-02 11:12:45,781] torch._dynamo.guards: [WARNING] Adding shape guard Eq(2*s2, s2 + 6)
[2023-03-02 11:12:46,066] torch._dynamo.guards: [WARNING] Specializing s0 to 7
[2023-03-02 11:12:46,067] torch._dynamo.guards: [WARNING] Specializing s1 to 8
```

All stack traces are moved under DEBUG.

ezyang commented:
What I mean by dogfooding is, don't just run the logging on the example, try to use the logging to diagnose the problem (pretend you don't know the answer to the issue). Is the logging giving you the info you need?

avikchaudhuri committed:
Update on "debug shape guards"






Adds logging when shape guards are added and when symbols are specialized to constants.

Differential Revision: [D43719743](https://our.internmc.facebook.com/intern/diff/D43719743/)

Differential Revision: [D43719743](https://our.internmc.facebook.com/intern/diff/D43719743)

[ghstack-poisoned]

avikchaudhuri commented:
Used this to successfully debug https://github.com/pytorch/pytorch/issues/95868

Logs now look like the following (user stack):
```
[2023-03-10 15:30:23,673] torch._dynamo.guards: [WARNING] Adding shape guard x['input_ids'].size()[1] < self.bert.embeddings.position_ids._base.size()[0] at
  File "<snip>/scripts/avik/pt2/example.py", line 399, in forward
    def forward(self, x):
  File "<snip>/torch/nn/modules/module.py", line 1526, in _call_impl
    def _call_impl(self, *args, **kwargs):
  File "<snip>/transformers/models/bert/modeling_bert.py", line 912, in forward
    @add_start_docstrings_to_model_forward(BERT_INPUTS_DOCSTRING.format("batch_size, sequence_length"))
  File "<snip>/torch/nn/modules/module.py", line 1526, in _call_impl
    def _call_impl(self, *args, **kwargs):
  File "<snip>/transformers/models/bert/modeling_bert.py", line 201, in forward
    def forward(
  File "<snip>/transformers/models/bert/modeling_bert.py", line 217, in forward
    position_ids = self.position_ids[:, past_key_values_length : seq_length + past_key_values_length]

[2023-03-10 15:30:23,702] torch._dynamo.guards: [WARNING] Adding shape guard Eq(x['input_ids'].size()[1], x['token_type_ids'].size()[1]) at
  File "<snip>/scripts/avik/pt2/example.py", line 399, in forward
    def forward(self, x):
  File "<snip>/torch/nn/modules/module.py", line 1526, in _call_impl
    def _call_impl(self, *args, **kwargs):
  File "<snip>/transformers/models/bert/modeling_bert.py", line 912, in forward
    @add_start_docstrings_to_model_forward(BERT_INPUTS_DOCSTRING.format("batch_size, sequence_length"))
  File "<snip>/torch/nn/modules/module.py", line 1526, in _call_impl
    def _call_impl(self, *args, **kwargs):
  File "<snip>/transformers/models/bert/modeling_bert.py", line 201, in forward
    def forward(
  File "<snip>/transformers/models/bert/modeling_bert.py", line 234, in forward
    embeddings = inputs_embeds + token_type_embeddings

[2023-03-10 15:30:23,932] torch._dynamo.guards: [WARNING] Adding shape guard Eq(x['input_ids'].size()[1], x['attention_mask'].size()[1]) at
  File "<snip>/scripts/avik/pt2/example.py", line 399, in forward
    def forward(self, x):
  File "<snip>/torch/nn/modules/module.py", line 1526, in _call_impl
    def _call_impl(self, *args, **kwargs):
  File "<snip>/transformers/models/bert/modeling_bert.py", line 912, in forward
    @add_start_docstrings_to_model_forward(BERT_INPUTS_DOCSTRING.format("batch_size, sequence_length"))
  File "<snip>/torch/nn/modules/module.py", line 1526, in _call_impl
    def _call_impl(self, *args, **kwargs):
  File "<snip>/transformers/models/bert/modeling_bert.py", line 562, in forward
    def forward(
  File "<snip>/torch/nn/modules/module.py", line 1526, in _call_impl
    def _call_impl(self, *args, **kwargs):
  File "<snip>/transformers/models/bert/modeling_bert.py", line 484, in forward
    def forward(
  File "<snip>/torch/nn/modules/module.py", line 1526, in _call_impl
    def _call_impl(self, *args, **kwargs):
  File "<snip>/transformers/models/bert/modeling_bert.py", line 416, in forward
    def forward(
  File "<snip>/torch/nn/modules/module.py", line 1526, in _call_impl
    def _call_impl(self, *args, **kwargs):
  File "<snip>/transformers/models/bert/modeling_bert.py", line 275, in forward
    def forward(
  File "<snip>/transformers/models/bert/modeling_bert.py", line 351, in forward
    attention_scores = attention_scores + attention_mask

```

avikchaudhuri commented:
For https://github.com/pytorch/pytorch/issues/93468:

```
[2023-03-10 15:35:04,007] torch._dynamo.guards: [WARNING] Adding shape guard Eq(lengths.size()[0], 10) at
  File "/data/users/avik/fbsource/buck-out/v2/gen/fbcode/4a9cc815958e3020/scripts/avik/pt2/__example__/example#link-tree/scripts/avik/pt2/example.py", line 338, in <resume in sequence_mask>
    torch.arange(0, max_len, device=lengths.device)
  File "/data/users/avik/fbsource/buck-out/v2/gen/fbcode/4a9cc815958e3020/scripts/avik/pt2/__example__/example#link-tree/scripts/avik/pt2/example.py", line 338, in <resume in sequence_mask>
    torch.arange(0, max_len, device=lengths.device)
```

avikchaudhuri committed:
Update on "debug shape guards"






Adds logging when shape guards are added and when symbols are specialized to constants.

Differential Revision: [D43719743](https://our.internmc.facebook.com/intern/diff/D43719743/)

Differential Revision: [D43719743](https://our.internmc.facebook.com/intern/diff/D43719743)

[ghstack-poisoned]

ezyang commented on torch/_guards.py:
```diff
@@ -341,6 +344,22 @@ def current_frame(frame_summary):
+        try:
+            yield
+        finally:
+            tc.loc_in_frame = None
```
I am a bit confused about the interaction between this and `frame_summary_stack`. Don't I want the current location in frame to be reported whenever I get the `frame_summary_stack`? I feel like the two of these should be interacting somehow.

avikchaudhuri replied:
The current implementation of `frame_summary_stack` gives a stack of function definition locs rather than function call locs. (In other words, `def foo(x) at ...` rather than `foo(t) at ...`. This doesn't quite look like a traditional stack trace.

ezyang replied:
OK, well, in a conventional stack trace, `def foo(x): ...` isn't included in a trace that throws midway through foo at all! So we probably don't want to do this?

ezyang replied:
Also, maybe we should just be using https://github.com/pytorch/pytorch/pull/95541 ?

avikchaudhuri replied:
OK, noting this as follow up work.

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -705,6 +715,12 @@ def safe_expand(r):
+
+def get_debugging_stack():
+    # cut this frame and the caller's frame
+    return ''.join(traceback.format_list(traceback.extract_stack()[:-2]))
```
nit: I find it helpful to take in the number of frames to remove as an argument to this function

avikchaudhuri replied:
done

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1803,6 +1824,18 @@ def _make_data_dependent_error(self, expr, unhinted_expr):
+            if torch._dynamo.config.print_specializations and isinstance(expr, (sympy.Integer, sympy.Float)):
+                # specializing to a constant, which is likely unexpected
+                torch._dynamo.guards.log.warning(f"Specializing {self.var_to_sources[a][0].name()} to {expr}")
+                torch._dynamo.guards.log.debug(f"Stack trace:\n{get_debugging_stack()}")
```
Instead of doing this by hand, you can just call `stack_info=True` to the logging call

avikchaudhuri replied:
done

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1803,6 +1824,18 @@ def _make_data_dependent_error(self, expr, unhinted_expr):
+        if a not in self.replacements or expr != self.replacements[a]:
+            if torch._dynamo.config.print_specializations and isinstance(expr, (sympy.Integer, sympy.Float)):
+                # specializing to a constant, which is likely unexpected
+                torch._dynamo.guards.log.warning(f"Specializing {self.var_to_sources[a][0].name()} to {expr}")
```
Call do torch._dynamo.guards log is not conventional; we're supposed to create a logger for this module like

```
log = logging.getLogger(__name__)
```

in fact, we already have one here.

avikchaudhuri replied:
I didn't see anything on the command line when I used the symbolic shapes log. Is it disabled?

ezyang replied:
I think the problem is that torch.fx.experimental is not in the same module namespace as torch._dynamo and torch._inductor, and the "PT2" logging only enables logs on certain parts of the module hierarchy. So the fix would be to enable it on torch.fx.experimental.symbolic_shapes, this seems fine.

avikchaudhuri replied:
done

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1816,7 +1849,7 @@ def _find(self, a: "sympy.Symbol") -> "sympy.Expr":
         res = self.replacements[a]
         cur_replace = {s: self._find(s) for s in res.free_symbols}
-        self.replacements[a] = self.replacements[a].xreplace(cur_replace)
+        self._set_replacement(a, self.replacements[a].xreplace(cur_replace))
```
would like to check if the extra equality test for every union find replacement is a performance problem

avikchaudhuri replied:
I put it under the condition for printing, which includes a constant check. Couldn't get rid of it entirely (otherwise will get a duplicate log message); added a comment.

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1885,6 +1918,25 @@ def _simplify_floor_div(self, expr):
+    def _add_guard(self, expr: "sympy.Expr") -> None:
+        stack = get_debugging_stack()
+        guard = ShapeGuard(expr, stack)
+        if os.environ.get("TORCHDYNAMO_PRINT_GUARDS", None) == "1":
```
We shouldn't be repeatedly querying the environment every time we add a guard

avikchaudhuri replied:
fixed

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1927,18 +1979,13 @@ def evaluate_expr(self, expr: "sympy.Expr", hint=None):
             # maybe_guard_eq in those cases.
             self._maybe_guard_eq(sympy.Eq(expr, concrete_val), True)
 
-        # TODO: optimize this; avoid formatting traces until we need them
```
This TODO is still valid

avikchaudhuri replied:
done

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1885,6 +1918,25 @@ def _simplify_floor_div(self, expr):
+            frame_summaries = TracingContext.get().frame_summary_stack
+            # frame_summaries describes a stack of functions
+            # TODO(avik): It would be better to describe a stack of function calls instead
+            current_loc = TracingContext.get().loc_in_frame
```
Yeah, so I feel like there should be a helper that gets this all in one go.

Also, are you sure the topmost frame isn't expressed twice?

avikchaudhuri replied:
See response above.

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1885,6 +1918,25 @@ def _simplify_floor_div(self, expr):
+            try:
+                expr = LoggingShapeGuardPrinter(self.var_to_sources).doprint(expr)
+            except Exception:
+                pass
```
This seems bad. At least log the exception!!

avikchaudhuri replied:
fixed

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1885,6 +1918,25 @@ def _simplify_floor_div(self, expr):
+                expr = LoggingShapeGuardPrinter(self.var_to_sources).doprint(expr)
+            except Exception:
+                pass
+            torch._dynamo.guards.log.warning(f"Adding shape guard {expr} at \n{user_stack}")
```
If try-except failed, expr won't be bound. I just wouldn't do a try-except at all.

ezyang reviewed (changes requested):
This broadly looks good but the bare try-except cannot go in, I'm happy to approve after that is fixed.

avikchaudhuri committed:
Update on "debug shape guards"






Adds logging when shape guards are added and when symbols are specialized to constants.

Differential Revision: [D43719743](https://our.internmc.facebook.com/intern/diff/D43719743/)

Differential Revision: [D43719743](https://our.internmc.facebook.com/intern/diff/D43719743)

[ghstack-poisoned]

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1885,6 +1918,23 @@ def _simplify_floor_div(self, expr):
+            # TODO: optimize this; avoid formatting traces until we need them
+            user_stack = ''.join(traceback.format_list([*frame_summaries, current_loc]))
+            expr = LoggingShapeGuardPrinter(self.var_to_sources).doprint(expr)
+            torch._dynamo.guards.log.warning(f"Adding shape guard {expr} at \n{user_stack}")
```
I think the easiest way to avoid useless string formatting when it is not necessary is to test the log level and only do the string formatting message when it is high enough.

avikchaudhuri replied:
done

avikchaudhuri committed:
Update on "debug shape guards"






Adds logging when shape guards are added and when symbols are specialized to constants.

Differential Revision: [D43719743](https://our.internmc.facebook.com/intern/diff/D43719743/)

Differential Revision: [D43719743](https://our.internmc.facebook.com/intern/diff/D43719743)

[ghstack-poisoned]

avikchaudhuri commented:
@pytorchbot merge -f "skip flaky ci"
